Improve pull logic, refactor cache

stelabouras opened the pull request:
The `fetchTranslations` method now offers a completion handler so that
the caller is notified when the downloading of the translations is
complete.

The completion block of that method now also reports any errors that
may have been encountered during the download operation so that the
caller can decide how the resulting translations should be treated.

`LocaleStrings` and all of the depended type aliases have now a `TX`
prefix to avoid any unwanted naming collisions.

Caching logic has been improved: If an object that conforms to the `TXCache`
protocol (which has also been renamed) isn't provided during the
`TxNative` initialization, then the internal `MemoryCache` is used,
which during its initialization looks up both in the main app bundle and
in the application directory to find any already generated or downloaded
translation files. If those files are found, the memory cache is updated
accordingly.

Also, when the new translations are fetched from CDS, the downloaded file
is either generated or replaced by those new translations.

The developers can now choose whether cache is used at all by using the
`useCache` argument of the `TxNative` initializer, which defaults to true.

Also, for applications that make use of several different targets (e.g.
extensions), the `TxNative` offers a new argument (`groupIdentifier`) so
that the downloaded translations are stored in the application group
directory in order for all of the targets to be able to have access to
them.

dbendilas commented:
I went through the code and I think the architecture looks somewhat different to what I had in mind. My concern is that the separation of concerns will not work very well as is, and that it's not making it as easy for developers to extend it as it could.

The following diagram shows better what I have in mind:
![TX Native Mobile SDK Cache](https://user-images.githubusercontent.com/22858669/102275201-bce2f200-3f2d-11eb-87bb-841f7ee582e1.png)

Initialization:
```swift
let downloadURL: URL = ...
let bundleURL: URL = ...
self.cache = cache ?? FileUpdateDecorator(
    fileURL: downloadURL, internalCache: ProviderBasedCache(
        providers: [
            DiskCacheProvider(fileURL: bundleURL),
            DiskCacheProvider(fileURL: downloadURL)
        ],
        replaceEntries: false,
        internalCache: MemoryCache()
    )
)
```

This way, we also don't need `useCache` and `groupIdentifier`, which didn't feel right to me for a high level interface as the initialization of `TXNative`; these are low-level internals of the disk loading mechanism and shouldn't be part of the initialization of the whole SDK. For the tool usage, we could provide a simple `MemoryCache` instance so that `useCache` is not needed (or even introduce a `NoopCache` if necessary -not sure what the actual needs are).

Since the above architecture is a bit complex, and since users might need to provide a custom cache in order to support extensions, maybe we could encapsulate this whole business logic in a new class, e.g.

```swift
TXNative.initialize(
    ...,
    cache: DefaultCache(groupIdentifier: "...")
)
```

We probably need to discuss this, as I might be missing something, or in order to clarify my intentions in a better way.

stelabouras commented:
> I went through the code and I think the architecture looks somewhat different to what I had in mind. My concern is that the separation of concerns will not work very well as is, and that it's not making it as easy for developers to extend it as we could.
> 
> The following diagram shows better what I have in mind:
> 
> ![TX Native Mobile SDK Cache](https://user-images.githubusercontent.com/22858669/102275201-bce2f200-3f2d-11eb-87bb-841f7ee582e1.png)
> 
> ```swift
> // NativeCore.init
> let downloadURL: URL = ...
> let bundleURL: URL = ...
> self.cache = cache ?? FileUpdateDecorator(
>     fileURL: downloadURL, 
>     internalCache: ProviderBasedCache(
>         providers: [
>             DiskCacheProvider(fileURL: bundleURL),
>             DiskCacheProvider(fileURL: downloadURL)
>         ],
>         replaceEntries: false,
>         internalCache: MemoryCache()
>     )
> )
> ```
> 
> This way, we also don't need `useCache` and `groupIdentifier`, which didn't feel right to me for a high level interface as the initialization of `TXNative`; these are low-level internals of the disk loading mechanism and shouldn't be part of the initialization of the whole SDK. For the tool usage, we could provide a simple `MemoryCache` instance so that `useCache` is not needed (or even introduce a `NoopCache` if necessary -not sure what the actual needs are).
> 
> Since the above architecture is a bit complex, and since users might need to provide a custom cache in order to support extensions, maybe we could encapsulate this whole business logic in a new class, e.g.
> 
> ```swift
> TXNative.initialize(
>     ...,
>     cache: DefaultCache(groupIdentifier: "...")
> )
> ```
> 
> We probably need to discuss this, as I might be missing something, or in order to clarify my intentions in a better way.

Just updated the branch with the architecture we discussed. Let me know what you think.

dbendilas reviewed (commented):
Looks great overall :+1: :+1:

I have created an updated diagram to better illustrate 1-2 points that I tried to describe with my previous suggestion, and also add a few more that came to me now (sorry I didn't have these in advance, they came to me after seeing the updates).

![TX Native Mobile SDK Cache](https://user-images.githubusercontent.com/22858669/102830978-53625800-43f3-11eb-868e-48db1446b5fe.png)


1. First of all, I have introduced a `TXDecoratorCache` class for convenience, as there are many classes that use the `internalCache` mechanism.
1. I believe that the behavior "save cache to disk" should be encapsulated into a class, separating concerns and not be bound to `TXDefaultCache`. My suggestion is the `TXFileOutputCacheDecorator` you see in the diagram. It delegates both function calls (`get()` and `update()`) to its `internalCache` and in addition it overrides `update()` so that it also saves the cache contents to the disk. 
1. I also suggest the introduction of `TXProviderBasedCache` which steals functionality away from `TXDefaultCache`, giving a more specialized (and reusable) home to `bundledTranslationsURL` and `downloadedTranslationsURL`.
1. I think `replaceEntries: Bool` isn't enough to describe what happens with the replacing/non-replacing mechanism. Seeing the implementation that doesn't replace strings, I have trouble understanding exactly what it does; and I think that we might have a different interpretation on what it means. Based on that, I can see developers wanting different things when treating cache update, so I have introduced `TXStringOverrideFilterCache` and `TXCacheOverridePolicy`. I have also removed `replaceEntries: Bool` from the signature of `TXCache` (and all its implementors), as I find it unnecessary to be a part of the signature; Instead, the same behavior can be achieved by the `TXStringOverrideFilterCache` decorator.
1. I suggest `TXStandardCache` instead of `TXDefaultCache`, although this is a minor change.
1. In my mind, `TXStandardCache` is a simple wrapper that uses the other decorators and classes to achieve the most common scenario(s) of caches that I foresee being used. It is initialized through `init(overridePolicy: TXCacheOverridePolicy? = .overrideAll, groupIdentifier: String?)` and in its constructor it simply creating a composite `internalCache` instance that is used all over.
1. Regarding the functionality "don't update the memory cache when translations are downloaded", I also believe it should be encapsulated into a class, therefore I have introduced `TXReadonlyCacheDecorator`. It overrides `update()`, making it no-op, while delegating the `get()` calls to `internalCache`.

---

The constructor of `TXStandardCache` could be something like:

```swift
cache = TXStringOverrideFilterCache(
    policy: overridePolicy,
    internalCache: TXFileOutputCacheDecorator(
        fileURL: downloadURL,
        internalCache: TXReadonlyCacheDecorator(
            TXProviderBasedCache(
                providers: [
                    TXBundleCacheProvider(fileURL: bundleURL, groupIdentifier: groupIdentifier),
                    TXDiskCacheProvider(fileURL: downloadURL)
                ],
                internalCache: MemoryCache()
            )
        )
    )
)
```

stelabouras commented:
> Looks great overall 👍 👍

I have begun implementing the above suggestions and to be honest, the number of different classes and how the inter-operate with each other can get quite overwhelming. It's a bit difficult to wrap my head around the decorator pattern in such an extensive way but I will try to follow your trail of thought here.

One of my concerns in terms of the suggested implementation is the use of the `groupIdentifier` argument and how it's passed down to the proper class (`TXBundleCacheIdentifier`) and which class is responsible for using it to turn it into a `fileURL` (for the `TXFileOutputCacheDecorator` to use it for storing the downloaded strings to the disk).

For starters, the "bundle cache" shouldn't care about the `groupIdentifier` as it just loads the translations from the app bundle (if they exist) and not from the sandbox directory (where the `groupIdentifier` is necessary so that it can look at the proper place). Given that, I have decided not to implement the `TXBundleCacheIdentifier` class.

Also, in the code snippet for the constructor of the `TXStandardCache`, it looks like the `downloadURL` and `bundleURL` are already constructed before being passed to the decorators, therefore I assume that the `downloadedTranslationsURL` and `bundledTranslationsURL` static methods continue to exist in the `TXStandardCache` (ex-`TXDefaultCache`) class. So for now, I have decided to leave the methods that generate the `downloadURL` and `bundleURL` in the `TXStandardCache` as both the providers and the `TXFileOutputCacheDecorator` need them.

The other concern I have is the override policy values. Let me know if the logic below is the one you want to be implemented:

* `.overrideAll`: Just replaces the whole cache with the new translations.

* `.overrideUntranslatedOnly`: Iterates through the cache and only updates a translation if a string key for a certain locale isn't found in the cache. If a string key for a certain locale is found in the cache but not in the updated translations, it's left untouched.

* `.overrideUsingTranslatedOnly`: Iterates through the cache and only updates a translation if a string key for a certain locale is found in the cache. If a string key for a certain locale is not found in the cache but it exists in the updated translations, it's not added in the cache. If a string key for a certain locale is found in the cache but not in the updated translations, it's left untouched.

I am pretty confident for the first two values of the override policy, but not so much for the last one (more on that below), so let me know what you think.

Also, I think that the override policy it's not usable the way we want it to be, as the updates should affect how the providers update the memory cache during the initialization of the `TXProviderBasedCache` and not how the downloaded strings are going to affect the data written in the downloaded file url.

So instead of the suggested code snippet that you posted, based on my previous sentence, I propose the following reordering of calls:

```swift
cache = TXFileOutputCacheDecorator(
    fileURL: downloadURL,
    internalCache: TXFileOutputCacheDecorator(
        fileURL: downloadURL,
        internalCache: TXReadonlyCacheDecorator(
            internalCache: TXProviderBasedCache(
                providers: providers,
                internalCache: TXStringOverrideFilterCache(
                    policy: overridePolicy,
                    internalCache: TXMemoryCache()
                )
            )
        )
    )
)
```

Given the above and my understanding of the override policy, it doesn't look like the `.overrideUsingTranslatedOnly` makes a lot of sense, as -given an empty cache- it would never update it as there are no translated strings in the beginning.

Maybe I got this all wrong though, so let me know.

All of the above changes have been implemented, I will just wait for your comments so that I can finalize the implementation.

stelabouras commented:
@dbendilas I have just pushed the final commit regarding the cache refactor as per our last conversation. I have also updated the README as discussed and updated the respective unit tests.

Take a look whenever you get the chance and let me know what you think.

stelabouras commented on Sources/TransifexNative/Cache.swift:
```diff
@@ -42,35 +43,303 @@ public protocol Cache {
+/// in the order they are added in the providers list.
+@objc
+public final class TXProviderBasedCache: NSObject {
+    let memoryCache: TXCache
```
Done

stelabouras commented on Sources/TransifexNative/Cache.swift:
```diff
@@ -42,35 +43,303 @@ public protocol Cache {
-    public override init() {
-        self.translationsByLocale = [:]
+    public init(groupIdentifier: String? = nil,
+                replaceEntries: Bool = false) {
```
`replaceEntries` argument has been removed

stelabouras commented on Sources/TransifexNative/Cache.swift:
```diff
@@ -42,35 +43,303 @@ public protocol Cache {
+}
+
+@objc
+public final class TXDefaultCache: NSObject {
```
Documentation added for all classes.

stelabouras commented on Sources/TransifexNative/Cache.swift:
```diff
@@ -42,35 +43,303 @@ public protocol Cache {
+                                                  replaceEntries: replaceEntries)
+    }
+    
+    /// - Returns: The URL of the translations file in the main bundle of the app
```
Added a more detailed docstring here.

stelabouras commented on Sources/TransifexNative/Cache.swift:
```diff
@@ -42,35 +43,303 @@ public protocol Cache {
+
+/// Composite class that accepts a number of cache providers, an internal cache and whether the providers
+/// should replace the entries of the cache or not. The providers are then used to update the internal class
+/// in the order they are added in the providers list.
```
Class has been refactored and documented.

dbendilas commented on Sources/TransifexNative/Cache.swift:
```diff
@@ -103,73 +119,257 @@ public final class TXDiskCacheProvider: NSObject, TXCacheProvider {
+                    continue
+                }
+                
+                    // If the policy set set to override untranslated only, then
```
`set set`

stelabouras replied:
Fixed

dbendilas commented on Sources/TransifexNative/Cache.swift:
```diff
@@ -103,73 +119,257 @@ public final class TXDiskCacheProvider: NSObject, TXCacheProvider {
+                        && self.get(key: stringKey,
+                                localeCode: localeCode) == nil)
+                   ||
+                    // If the policy is set to override using translated only,
+                    // then update the cache only if there's an existing
+                    // translation for that stringKey.
+                    (policy == .overrideUsingTranslatedOnly
+                        && self.get(key: stringKey,
+                                    localeCode: localeCode) != nil) {
```
I think this is not correct. `overrideUsingTranslatedOnly` for me means that we'll never replace a non-empty translation in the internal cache with an empty translation found in the incoming translations.

You are already setting `translatedString` with a `guard`, so we know that it's not empty. Therefore, the second conditional should simply be `policy == .overrideUsingTranslatedOnly`.

stelabouras replied:
My bad, I thought the `overrideUsingTranslatedOnly` policy meant that the string key should also be found in the cache in order to be updated. I will update the logic there by changing the second conditional as described.

I will also update the docstring for the `overrideUsingTranslatedOnly` value to be:

```
    /// All new translations are added to the cache, either updating existing translations or adding
    /// new ones.
    /// If a translation is not found in cache but exists in the new translations, it's not added.
    /// If a translation is found in cache but doesn't exist in the new translations, it's left untouched.
    /// Empty cache entries from the new translations are filtered out.
```

dbendilas commented on Sources/TransifexNative/Cache.swift:
```diff
@@ -103,73 +119,257 @@ public final class TXDiskCacheProvider: NSObject, TXCacheProvider {
     }
 }
 
+/// The standard wrapper cache that the TxNative SDK is initialized with, if no other cache is provided.
```
I would remove `wrapper` from here, I don't think it provides any useful context. I would go with `The default cache` or `The standard cache`.

stelabouras replied:
Fixed

dbendilas reviewed (changes requested):
Great job! Very nice implementation and descriptive docstrings 👍 

There are a few things I've commented on, almost all are minor. One additional thing is that I'd like a test for the readonly cache, to make sure it behaves as advertised.

stelabouras commented on Sources/TransifexNative/Cache.swift:
```diff
@@ -103,73 +119,257 @@ public final class TXDiskCacheProvider: NSObject, TXCacheProvider {
+        guard let serializedData = jsonData,
+              let serializedTranslations = String(data: serializedData,
+                                                  encoding: .utf8) else {
+            throw StoringErrors.encodingFailed
```
Yes, the caller of this method (`update()` method of `TXFileOutputCacheDecorator`) has a proper `do { try } catch { }` block wrapped around that call to properly catch and handle any exceptions raised by the `storeTranslations()` method.

stelabouras commented on README.md:
```diff
@@ -125,6 +125,83 @@ to push them to CDS. You can optionally set the `purge` argument to `true` (defa
+
+Whenever new translations are fetched from the server using the `fetchTranslations()` 
+method, the standard cache is updated and those translations are stored as-is in the file 
+system, in the same cache fileused by the aforementioned second cache provider so that 
```
Fixed

stelabouras commented on Sources/TransifexNative/Cache.swift:
```diff
@@ -103,73 +119,257 @@ public final class TXDiskCacheProvider: NSObject, TXCacheProvider {
+/// let cache = TXProviderBasedCache(
+///     providers: [
+///         TXDiskCacheProvider(fileURL: firstFileURL),
+///         TXDiskCacheProvider(fileURL: secondFileURL
```
Fixed

stelabouras commented on Sources/TransifexNative/Cache.swift:
```diff
@@ -103,73 +119,257 @@ public final class TXDiskCacheProvider: NSObject, TXCacheProvider {
-    public func get(key: String, localeCode: String) -> String? {
-        return memoryCache.get(key: key, localeCode: localeCode)
+/// Decorator class responsible for updating the passed internalCache using a certain override policy defined
+/// in the `TXCacheOverridePolicy` enum.
```
Updated docstring with a more detailed explanation.

stelabouras commented:
@dbendilas Just added a Github workflow so that tests can run using Github Actions: https://github.com/transifex/transifex-swift/actions/runs/482411344

dbendilas reviewed (approved):
Everything looks good! If I'm not missing anything you know of, we can remove the `WIP` and merge.

I didn't know that it was so easy to run the test suite with GitHub Actions and I was planning on asking you to integrate some sort of CI for that. Big kudos!
